torch/_dynamo/output_graph.py

```diff
@@ -1863,6 +1863,9 @@ def create_proxy(
         if self.parent is not None:
             flat_args, tree_spec = pytree.tree_flatten((args, kwargs))
             new_flat_args = []
+            for arg in flat_args:
+                if isinstance(arg, torch.fx.proxy.Proxy):
+                    self.graph._graph_namespace._used_names.add(arg.node.name)
```

Comment:
Er, is the problem just that we get a name collision?
```
        contiguous_1 = contiguous
        contiguous_2 = contiguous_1
```
The problem sounds like FX's naming logic didn't realize that `contiguous_1` was already used. Would it be possible to solve the problem in torch.fx instead of here (which is for Dynamo)?

Reply:
> Er, is the problem just that we get a name collision?

Yes, exactly.

> Would it be possible to solve the problem in torch.fx instead of here (which is for Dynamo)?

Let me try to explain what happens here to better answer your question.

In the example above, the lifting happens when processing this function call
```python
# Let's call this node2. Note that it is not yet generated here.
 somefunc_backward_default: "f32[7]" = torch.ops.mylib.somefunc_backward.default(contiguous, contiguous_1, contiguous_2, [7]);
```
And the graph at this moment looks like
```python
graph():
    %function_ctx : [num_users=0] = placeholder[target=function_ctx]
    %somefunc_forward_default : [num_users=1] = placeholder[target=somefunc_forward_default]
    %_set_grad_enabled : [num_users=0] = call_function[target=torch._C._set_grad_enabled](args = (False,), kwargs = {})
    # Let's call this node1
    %contiguous : [num_users=0] = call_method[target=contiguous](args = (%somefunc_forward_default,), kwargs = {})
```
For the first three args of `node2`, the names of their proxies are `contiguous`, `contiguous` and `contiguous_1` respectively, where the last two directly come from the saved ctx in forward pass.

Thus, when trying to lift the second (`contiguous`), FX has no idea about the existence of `contiguous_1`. And thus Dyanmo creates a placeholder node, where the target is the `contiguous` proxy from forward.
```python
# Let's call it node3
%contiguous_1 : [num_users=0] = placeholder[target=contiguous]
```
Similarly, Dynamo creates a placeholder node for the third arg:
```python
# Let's call it node4
%contiguous_2 : [num_users=0] = placeholder[target=contiguous_1]
```

Thus, `node4` actually uses value from `node3` (`contiguous_1` and  `contiguous_2` actually become same variable) instead of the `contiguous_1` that was lifted to the input args.

As a conclusion, I feel it is hard to do this in FX side because it just doesn't know the args list. But I am new to Dynamo, I may miss something. Feel free to correct me.

Reply:
Let me try to summarize the problem:
- when we lift a freevar, we use the name of the freevar to create a graph input
- we create the graph input by creating a new FX node with kind="placeholder" and target=name
- it is possible that the target conflicts with some node that already exists in the FX graph.

If that sounds right, I'm not sure the fix in this PR is sufficient. Can you instead put a check in https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/_dynamo/output_graph.py#L2102 to see if the name of the new graph input is already taken, and if it is, then we generate a new unique name?

Reply:
For the second point, besides kind and target, the proxy name itself will be generated by FX in
https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/fx/graph.py#L1008
So for the third point, it is not the target conflicting with some nodes that already exists in the FX graph (the target just exists in the FX graph for forward pass, and we are currently generating for backward pass). It is more like the generated node name conflicting with some targets from previous tracer. 

In our case for `node3`, target="contiguous" and `node3`'s name is "contiguous_1" (generated by FX, because the first non-conflicting name for format "contiguous_{num}" is "contiguous_1"). And then when generating `node4`, the target of `node4` is "contiguous_1", which conflicts with `node3`'s name.


Btw, the "previous tracer" in our case is the the `fwd_tracer` here:
https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/_dynamo/variables/higher_order_ops.py#L1656

> Can you instead put a check in...

Thus, putting a check here cannot catch the issue, because "contiguous_1" is not in the FX graph when generating `node3`.
Note that given "contiguous_1" is in the previous FX graph, we may also enforce the FX graph for [bwd_tracer](https://github.com/pytorch/pytorch/blob/f9bb258892cc71b863d8eb3686a361150be07ee2/torch/_dynamo/variables/higher_order_ops.py#L1695) to consider the names of `fwd_tracer`. However, this still need to directly access private variable of FX graph.

Reply:
> Thus, putting a check here cannot catch the issue, because "contiguous_1" is not in the FX graph when generating node3.

Will a check here catch the issue when generating node4? When generating node4, we can pick a different target (so, instead of "contiguous_1" we could have picked "contiguous_1_1")

Reply:
Ah I see what you mean. Yes, this can catch the issue when generating node4. I pushed the fix.

Reply:
Instead of original manually finding unique input name, I replace that part of code with graph_namespace.create_name. PTAL.